File: test-suite/ruby/archive_extraction_clean/zip_slip_safe.rb
require "fileutils"
require "zip"
require "rubygems/package"

def safe_archive_path(destination, entry_name)
  base = File.expand_path(destination)
  target = File.expand_path(entry_name, base)
  unless target.start_with?(base + File::SEPARATOR) || target == base
    raise ArgumentError, "archive entry escapes destination"
  end
  target
end

def unzip_safely(archive_path, destination)
  Zip::File.open(archive_path) do |zip_file|
    zip_file.each do |entry|
      target = safe_archive_path(destination, entry.name)
      FileUtils.mkdir_p(File.dirname(target))
      entry.extract(target) { true }
    end
  end
end

def untar_safely(io, destination)
  base = File.expand_path(destination)
  Gem::Package::TarReader.new(io).each do |entry|
    target = File.expand_path(entry.full_name, base)
    unless target.start_with?(base + File::SEPARATOR) || target == base
      raise ArgumentError, "archive entry escapes destination"
    end
    FileUtils.mkdir_p(File.dirname(target))
    File.binwrite(target, entry.read)
  end
end

File: test-suite/kotlin/archive_extraction_clean/zip_slip_safe.kt
import java.io.File
import java.nio.file.Files
import java.nio.file.Path
import java.util.zip.ZipFile
import java.util.zip.ZipInputStream

private fun safeArchivePath(destination: Path, entryName: String): Path {
    val base = destination.toAbsolutePath().normalize()
    val target = base.resolve(entryName).normalize()
    require(target.startsWith(base)) { "archive entry escapes destination" }
    return target
}

fun unzipWithZipFileSafely(archive: File, destination: Path) {
    ZipFile(archive).use { zip ->
        zip.entries().asSequence().forEach { entry ->
            val target = safeArchivePath(destination, entry.name)
            if (entry.isDirectory) {
                Files.createDirectories(target)
            } else {
                Files.createDirectories(target.parent)
                zip.getInputStream(entry).use { input ->
                    Files.copy(input, target)
                }
            }
        }
    }
}

fun unzipInlineSafely(stream: ZipInputStream, destination: File) {
    val base = destination.toPath().toAbsolutePath().normalize()
    generateSequence { stream.nextEntry }.forEach { entry ->
        val target = base.resolve(entry.name).normalize()
        require(target.startsWith(base)) { "archive entry escapes destination" }
        Files.copy(stream, target)
    }
}
